File: ts_gui/src/invoke/logger.ts
import { invoke } from '@tauri-apps/api/tauri';

// Function to log messages
export default class logger {
    public static async logMessage(level: string, message: string) {
        await invoke("plugin:logger|log_message", { level, message })
            .then(() => {
                console.log(`Log [${level}] sent to Rust backend successfully.`);
            })
            .catch((error) => {
                console.error(`Failed to send log [${level}] to Rust backend:`, error);
            });
    }

    public static async invokeWithLogging<T>(command: string, args?: Record<string, unknown>): Promise<T> {
        try {
            logger.logMessage("debug", `Invoking command: ${command} with args: ${JSON.stringify(args)}`);
            const result = await invoke<T>(command, args);
            logger.logMessage("info", `Command ${command} executed successfully.`);
            return result;
        } catch (error) {
            if (error instanceof Error) {
                logger.logMessage("error", `Error in command ${command}: ${error.message}`);
            } else {
                logger.logMessage("error", `Error in command ${command}: ${JSON.stringify(error)} `);
            }
            throw error; 
        }
    }
}

File: ts_gui/src-tauri/src/api/logger.rs
use tracing::{info, warn, error, debug};
use tauri::plugin::TauriPlugin;

pub fn init<R: tauri::Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::new("logger")
        .invoke_handler(tauri::generate_handler![
            log_message
        ])
        .build()
}

/// await invoke("plugin:logger|log_message", { level, message });
#[tauri::command]
pub async fn log_message(level: String, message: String) {
    match level.as_str() {
        "info" => info!("{}", message),
        "warn" => warn!("{}", message),
        "error" => error!("{}", message),
        "debug" => debug!("{}", message),
        _ => info!("{}", message),
    }
}